File: go/internal/e2e/mcp_and_agents_test.go
package e2e

import (
	"path/filepath"
	"strings"
	"testing"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
)

func TestMCPServers(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	client := ctx.NewClient()
	t.Cleanup(func() { client.ForceStop() })

	t.Run("accept MCP server config on create", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		mcpServers := map[string]copilot.MCPServerConfig{
			"test-server": {
				"type":    "local",
				"command": "echo",
				"args":    []string{"hello"},
				"tools":   []string{"*"},
			},
		}

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			MCPServers: mcpServers,
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		if session.SessionID == "" {
			t.Error("Expected non-empty session ID")
		}

		// Simple interaction to verify session works
		_, err = session.Send(t.Context(), copilot.MessageOptions{
			Prompt: "What is 2+2?",
		})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		message, err := testharness.GetFinalAssistantMessage(t.Context(), session)
		if err != nil {
			t.Fatalf("Failed to get final message: %v", err)
		}

		if message.Data.Content == nil || !strings.Contains(*message.Data.Content, "4") {
			t.Errorf("Expected message to contain '4', got: %v", message.Data.Content)
		}

		session.Destroy()
	})

	t.Run("accept MCP server config on resume", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		// Create a session first
		session1, err := client.CreateSession(t.Context(), nil)
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}
		sessionID := session1.SessionID

		_, err = session1.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: "What is 1+1?"})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		// Resume with MCP servers
		mcpServers := map[string]copilot.MCPServerConfig{
			"test-server": {
				"type":    "local",
				"command": "echo",
				"args":    []string{"hello"},
				"tools":   []string{"*"},
			},
		}

		session2, err := client.ResumeSessionWithOptions(t.Context(), sessionID, &copilot.ResumeSessionConfig{
			MCPServers: mcpServers,
		})
		if err != nil {
			t.Fatalf("Failed to resume session: %v", err)
		}

		if session2.SessionID != sessionID {
			t.Errorf("Expected session ID %s, got %s", sessionID, session2.SessionID)
		}

		message, err := session2.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: "What is 3+3?"})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		if message.Data.Content == nil || !strings.Contains(*message.Data.Content, "6") {
			t.Errorf("Expected message to contain '6', got: %v", message.Data.Content)
		}

		session2.Destroy()
	})

	t.Run("should pass literal env values to MCP server subprocess", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		mcpServerPath, err := filepath.Abs("../../../test/harness/test-mcp-server.mjs")
		if err != nil {
			t.Fatalf("Failed to resolve test-mcp-server path: %v", err)
		}
		mcpServerDir := filepath.Dir(mcpServerPath)

		mcpServers := map[string]copilot.MCPServerConfig{
			"env-echo": {
				"type":    "local",
				"command": "node",
				"args":    []string{mcpServerPath},
				"tools":   []string{"*"},
				"env":     map[string]string{"TEST_SECRET": "hunter2"},
				"cwd":     mcpServerDir,
			},
		}

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			MCPServers:          mcpServers,
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		if session.SessionID == "" {
			t.Error("Expected non-empty session ID")
		}

		message, err := session.SendAndWait(t.Context(), copilot.MessageOptions{
			Prompt: "Use the env-echo/get_env tool to read the TEST_SECRET environment variable. Reply with just the value, nothing else.",
		})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		if message.Data.Content == nil || !strings.Contains(*message.Data.Content, "hunter2") {
			t.Errorf("Expected message to contain 'hunter2', got: %v", message.Data.Content)
		}

		session.Destroy()
	})

	t.Run("handle multiple MCP servers", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		mcpServers := map[string]copilot.MCPServerConfig{
			"server1": {
				"type":    "local",
				"command": "echo",
				"args":    []string{"server1"},
				"tools":   []string{"*"},
			},
			"server2": {
				"type":    "local",
				"command": "echo",
				"args":    []string{"server2"},
				"tools":   []string{"*"},
			},
		}

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			MCPServers: mcpServers,
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		if session.SessionID == "" {
			t.Error("Expected non-empty session ID")
		}

		session.Destroy()
	})
}

func TestCustomAgents(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	client := ctx.NewClient()
	t.Cleanup(func() { client.ForceStop() })

	t.Run("accept custom agent config on create", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		infer := true
		customAgents := []copilot.CustomAgentConfig{
			{
				Name:        "test-agent",
				DisplayName: "Test Agent",
				Description: "A test agent for SDK testing",
				Prompt:      "You are a helpful test agent.",
				Infer:       &infer,
			},
		}

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			CustomAgents: customAgents,
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		if session.SessionID == "" {
			t.Error("Expected non-empty session ID")
		}

		// Simple interaction to verify session works
		_, err = session.Send(t.Context(), copilot.MessageOptions{
			Prompt: "What is 5+5?",
		})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		message, err := testharness.GetFinalAssistantMessage(t.Context(), session)
		if err != nil {
			t.Fatalf("Failed to get final message: %v", err)
		}

		if message.Data.Content == nil || !strings.Contains(*message.Data.Content, "10") {
			t.Errorf("Expected message to contain '10', got: %v", message.Data.Content)
		}

		session.Destroy()
	})

	t.Run("accept custom agent config on resume", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		// Create a session first
		session1, err := client.CreateSession(t.Context(), nil)
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}
		sessionID := session1.SessionID

		_, err = session1.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: "What is 1+1?"})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		// Resume with custom agents
		customAgents := []copilot.CustomAgentConfig{
			{
				Name:        "resume-agent",
				DisplayName: "Resume Agent",
				Description: "An agent added on resume",
				Prompt:      "You are a resume test agent.",
			},
		}

		session2, err := client.ResumeSessionWithOptions(t.Context(), sessionID, &copilot.ResumeSessionConfig{
			CustomAgents: customAgents,
		})
		if err != nil {
			t.Fatalf("Failed to resume session: %v", err)
		}

		if session2.SessionID != sessionID {
			t.Errorf("Expected session ID %s, got %s", sessionID, session2.SessionID)
		}

		message, err := session2.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: "What is 6+6?"})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		if message.Data.Content == nil || !strings.Contains(*message.Data.Content, "12") {
			t.Errorf("Expected message to contain '12', got: %v", message.Data.Content)
		}

		session2.Destroy()
	})

	t.Run("handle custom agent with tools", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		infer := true
		customAgents := []copilot.CustomAgentConfig{
			{
				Name:        "tool-agent",
				DisplayName: "Tool Agent",
				Description: "An agent with specific tools",
				Prompt:      "You are an agent with specific tools.",
				Tools:       []string{"bash", "edit"},
				Infer:       &infer,
			},
		}

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			CustomAgents: customAgents,
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		if session.SessionID == "" {
			t.Error("Expected non-empty session ID")
		}

		session.Destroy()
	})

	t.Run("handle custom agent with MCP servers", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		customAgents := []copilot.CustomAgentConfig{
			{
				Name:        "mcp-agent",
				DisplayName: "MCP Agent",
				Description: "An agent with its own MCP servers",
				Prompt:      "You are an agent with MCP servers.",
				MCPServers: map[string]copilot.MCPServerConfig{
					"agent-server": {
						"type":    "local",
						"command": "echo",
						"args":    []string{"agent-mcp"},
						"tools":   []string{"*"},
					},
				},
			},
		}

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			CustomAgents: customAgents,
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		if session.SessionID == "" {
			t.Error("Expected non-empty session ID")
		}

		session.Destroy()
	})

	t.Run("handle multiple custom agents", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		inferTrue := true
		inferFalse := false
		customAgents := []copilot.CustomAgentConfig{
			{
				Name:        "agent1",
				DisplayName: "Agent One",
				Description: "First agent",
				Prompt:      "You are agent one.",
				Infer:       &inferTrue,
			},
			{
				Name:        "agent2",
				DisplayName: "Agent Two",
				Description: "Second agent",
				Prompt:      "You are agent two.",
				Infer:       &inferFalse,
			},
		}

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			CustomAgents: customAgents,
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		if session.SessionID == "" {
			t.Error("Expected non-empty session ID")
		}

		session.Destroy()
	})
}

func TestCombinedConfiguration(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	client := ctx.NewClient()
	t.Cleanup(func() { client.ForceStop() })

	t.Run("accept MCP servers and custom agents", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		mcpServers := map[string]copilot.MCPServerConfig{
			"shared-server": {
				"type":    "local",
				"command": "echo",
				"args":    []string{"shared"},
				"tools":   []string{"*"},
			},
		}

		customAgents := []copilot.CustomAgentConfig{
			{
				Name:        "combined-agent",
				DisplayName: "Combined Agent",
				Description: "An agent using shared MCP servers",
				Prompt:      "You are a combined test agent.",
			},
		}

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			MCPServers:   mcpServers,
			CustomAgents: customAgents,
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		if session.SessionID == "" {
			t.Error("Expected non-empty session ID")
		}

		_, err = session.Send(t.Context(), copilot.MessageOptions{
			Prompt: "What is 7+7?",
		})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		message, err := testharness.GetFinalAssistantMessage(t.Context(), session)
		if err != nil {
			t.Fatalf("Failed to get final message: %v", err)
		}

		if message.Data.Content == nil || !strings.Contains(*message.Data.Content, "14") {
			t.Errorf("Expected message to contain '14', got: %v", message.Data.Content)
		}

		session.Destroy()
	})
}

File: nodejs/test/e2e/mcp_and_agents.test.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import { dirname, resolve } from "path";
import { fileURLToPath } from "url";
import { describe, expect, it } from "vitest";
import type { CustomAgentConfig, MCPLocalServerConfig, MCPServerConfig } from "../../src/index.js";
import { approveAll } from "../../src/index.js";
import { createSdkTestContext } from "./harness/sdkTestContext.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const TEST_MCP_SERVER = resolve(__dirname, "../../../test/harness/test-mcp-server.mjs");

describe("MCP Servers and Custom Agents", async () => {
    const { copilotClient: client } = await createSdkTestContext();

    describe("MCP Servers", () => {
        it("should accept MCP server configuration on session create", async () => {
            const mcpServers: Record<string, MCPServerConfig> = {
                "test-server": {
                    type: "local",
                    command: "echo",
                    args: ["hello"],
                    tools: ["*"],
                } as MCPLocalServerConfig,
            };

            const session = await client.createSession({
                mcpServers,
            });

            expect(session.sessionId).toBeDefined();

            // Simple interaction to verify session works
            const message = await session.sendAndWait({
                prompt: "What is 2+2?",
            });
            expect(message?.data.content).toContain("4");

            await session.destroy();
        });

        it("should accept MCP server configuration on session resume", async () => {
            // Create a session first
            const session1 = await client.createSession();
            const sessionId = session1.sessionId;
            await session1.sendAndWait({ prompt: "What is 1+1?" });

            // Resume with MCP servers
            const mcpServers: Record<string, MCPServerConfig> = {
                "test-server": {
                    type: "local",
                    command: "echo",
                    args: ["hello"],
                    tools: ["*"],
                } as MCPLocalServerConfig,
            };

            const session2 = await client.resumeSession(sessionId, {
                mcpServers,
            });

            expect(session2.sessionId).toBe(sessionId);

            const message = await session2.sendAndWait({
                prompt: "What is 3+3?",
            });
            expect(message?.data.content).toContain("6");

            await session2.destroy();
        });

        it("should handle multiple MCP servers", async () => {
            const mcpServers: Record<string, MCPServerConfig> = {
                server1: {
                    type: "local",
                    command: "echo",
                    args: ["server1"],
                    tools: ["*"],
                } as MCPLocalServerConfig,
                server2: {
                    type: "local",
                    command: "echo",
                    args: ["server2"],
                    tools: ["*"],
                } as MCPLocalServerConfig,
            };

            const session = await client.createSession({
                mcpServers,
            });

            expect(session.sessionId).toBeDefined();
            await session.destroy();
        });

        it("should pass literal env values to MCP server subprocess", async () => {
            const mcpServers: Record<string, MCPServerConfig> = {
                "env-echo": {
                    type: "local",
                    command: "node",
                    args: [TEST_MCP_SERVER],
                    tools: ["*"],
                    env: { TEST_SECRET: "hunter2" },
                } as MCPLocalServerConfig,
            };

            const session = await client.createSession({
                mcpServers,
                onPermissionRequest: approveAll,
            });

            expect(session.sessionId).toBeDefined();

            const message = await session.sendAndWait({
                prompt: "Use the env-echo/get_env tool to read the TEST_SECRET environment variable. Reply with just the value, nothing else.",
            });
            expect(message?.data.content).toContain("hunter2");

            await session.destroy();
        });
    });

    describe("Custom Agents", () => {
        it("should accept custom agent configuration on session create", async () => {
            const customAgents: CustomAgentConfig[] = [
                {
                    name: "test-agent",
                    displayName: "Test Agent",
                    description: "A test agent for SDK testing",
                    prompt: "You are a helpful test agent.",
                    infer: true,
                },
            ];

            const session = await client.createSession({
                customAgents,
            });

            expect(session.sessionId).toBeDefined();

            // Simple interaction to verify session works
            const message = await session.sendAndWait({
                prompt: "What is 5+5?",
            });
            expect(message?.data.content).toContain("10");

            await session.destroy();
        });

        it("should accept custom agent configuration on session resume", async () => {
            // Create a session first
            const session1 = await client.createSession();
            const sessionId = session1.sessionId;
            await session1.sendAndWait({ prompt: "What is 1+1?" });

            // Resume with custom agents
            const customAgents: CustomAgentConfig[] = [
                {
                    name: "resume-agent",
                    displayName: "Resume Agent",
                    description: "An agent added on resume",
                    prompt: "You are a resume test agent.",
                },
            ];

            const session2 = await client.resumeSession(sessionId, {
                customAgents,
            });

            expect(session2.sessionId).toBe(sessionId);

            const message = await session2.sendAndWait({
                prompt: "What is 6+6?",
            });
            expect(message?.data.content).toContain("12");

            await session2.destroy();
        });

        it("should handle custom agent with tools configuration", async () => {
            const customAgents: CustomAgentConfig[] = [
                {
                    name: "tool-agent",
                    displayName: "Tool Agent",
                    description: "An agent with specific tools",
                    prompt: "You are an agent with specific tools.",
                    tools: ["bash", "edit"],
                    infer: true,
                },
            ];

            const session = await client.createSession({
                customAgents,
            });

            expect(session.sessionId).toBeDefined();
            await session.destroy();
        });

        it("should handle custom agent with MCP servers", async () => {
            const customAgents: CustomAgentConfig[] = [
                {
                    name: "mcp-agent",
                    displayName: "MCP Agent",
                    description: "An agent with its own MCP servers",
                    prompt: "You are an agent with MCP servers.",
                    mcpServers: {
                        "agent-server": {
                            type: "local",
                            command: "echo",
                            args: ["agent-mcp"],
                            tools: ["*"],
                        } as MCPLocalServerConfig,
                    },
                },
            ];

            const session = await client.createSession({
                customAgents,
            });

            expect(session.sessionId).toBeDefined();
            await session.destroy();
        });

        it("should handle multiple custom agents", async () => {
            const customAgents: CustomAgentConfig[] = [
                {
                    name: "agent1",
                    displayName: "Agent One",
                    description: "First agent",
                    prompt: "You are agent one.",
                },
                {
                    name: "agent2",
                    displayName: "Agent Two",
                    description: "Second agent",
                    prompt: "You are agent two.",
                    infer: false,
                },
            ];

            const session = await client.createSession({
                customAgents,
            });

            expect(session.sessionId).toBeDefined();
            await session.destroy();
        });
    });

    describe("Combined Configuration", () => {
        it("should accept both MCP servers and custom agents", async () => {
            const mcpServers: Record<string, MCPServerConfig> = {
                "shared-server": {
                    type: "local",
                    command: "echo",
                    args: ["shared"],
                    tools: ["*"],
                } as MCPLocalServerConfig,
            };

            const customAgents: CustomAgentConfig[] = [
                {
                    name: "combined-agent",
                    displayName: "Combined Agent",
                    description: "An agent using shared MCP servers",
                    prompt: "You are a combined test agent.",
                },
            ];

            const session = await client.createSession({
                mcpServers,
                customAgents,
            });

            expect(session.sessionId).toBeDefined();

            const message = await session.sendAndWait({
                prompt: "What is 7+7?",
            });
            expect(message?.data.content).toContain("14");

            await session.destroy();
        });
    });
});
